Use fixed thread pool for deployments

hmusum opened the pull request:
We have seen 30 deployments running at the same time, which is more than the config servers can handle.

Commit:
Use fixed thread pool for deployments

hmusum commented:
Wait, I want to look into deploying another way

Commit:
Use a thread pool with 1 thread instead of throttling

baldersheim commented on node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/ApplicationMaintainer.java:
```diff
@@ -28,7 +27,10 @@
+    // Use a fixed thread pool to avoid overload on config servers. Resource usage when deploying varies
+    // a lot between applications, so doing one by one avoids issues where one or more resource-demanding
+    // deployments happen simultaneously
+    private final Executor deploymentExecutor = Executors.newSingleThreadExecutor(new DaemonThreadFactory("node repo application maintainer"));
```
Can we test if 1 is enough ?

hmusum replied:
Hm, we could try it out in one zone and try to gather some numbers, but I'll have to look into how to do that.

baldersheim commented on node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/ApplicationMaintainer.java:
```diff
@@ -41,7 +43,6 @@ protected final void maintain() {
         for (ApplicationId application : applications) {
```
Do we have logging outside of this loop or should we add it to see how long it takes ?

hmusum replied:
Since the deployments will just be queued up I don't think it will help us much, we would need to somehow wait until all submitted jobs were finished then

baldersheim replied:
Is it not possible to to do an executor.sync or something to test that it is done.

baldersheim reviewed (approved):
Consider comments, merge at will.

bratseth commented:
It's not important any more that it happens every 30 minutes, and we know when each application last deployed. We could just use a single thread which picks the oldest deployment and redeploys that.

hmusum commented:
Thanks for the suggestion, I'll look into it

Commit:
Deploy the app that was deployed the longest time ago

Run PeriodicApplicationMaintainer more often and deploy the app
that was deployed the longest time ago
